### ISSUE TITLE
[BUG] Not possible to set the bslamp2 to warm light

### ISSUE THREAD

**Issue by helgek** [mmakaay/esphome-yeelight_bs2#126]
**Describe the bug**
In comparision with the original firmware I'm not able to get a light color reproduction of 1700K in CCT mode.

**To Reproduce**
Activate CCT mode and chose Kelvin color setting in the range of 1700K-2000K.

**Expected behavior**
Warm light recognisably representing the 1700K-2000K range.

**Additional context**
Please see the following comparison of ESPHome firmware and original firmware in pictures. 

You can see that the ESPHome based nightlight works really well, it's even darker than the one with the original firmware. At 2% the ESPHome based is brighter than the one with original firmware. But this could be due to the wrong representation of light color and if it would correctly do the 1700K the visible brightness might be similar.

![IMG20240310032412](https://github.com/mmakaay/esphome-xiaomi_bslamp2/assets/6280991/cf89b111-f44d-4864-b5cf-673174b57255)

![IMG20240310032445](https://github.com/mmakaay/esphome-xiaomi_bslamp2/assets/6280991/318cb9c0-5c67-471e-a15a-c94e20352d39)

![IMG20240310032755](https://github.com/mmakaay/esphome-xiaomi_bslamp2/assets/6280991/dd069731-34b5-492a-a65c-9077c34f7651)

**Comment by mmakaay** [comment]
What I've seen with other devices too, is that there seems to be some calibration required on the hardware in order to get the same results as the original firmware. The ways the outputs are driven now are derived from me measuring at many (MANY!) different light settings, and building lookup tables out of those to find and interpolate settings. Unfortunately, the output levels looked nothing like clean linear RGBWW output values.

Since there are more reports of light outputs not matching the original outputs, I think that some additional parameters for calibrating the light are required. For example a scaling factor for every channel, that linearly changes the output level for those channels. That will require some experimentation to see if that is a feasible work-around here. I will look into it to see if I can come up with something.

For the night light output levels, I know that some people tweaked the output levels in code to get their desired output. Even to tweak the night light color to a desired color, e.g. blue. That would probably not work for your case, since you also have issues with the non-nightlight settings. Updating the whole output value table is not something I would recommend. It was painful enough, going from measurements to a table. Since the relation between output levels and resulting light color are a bit weird, changing values to get the desired light output is likely quite the ordeal.

**Comment by vojtechj** [comment]
wouldn't it be possible to get original light values from extracted original firmware?

**Comment by vojtechj** [comment]
and could this issue be related to the problem that if I set yellow color in home assistant the lamp outputs something between yellow and green .... other colors look good.

**Comment by helgek** [comment]
[fwbackup_bedside_lamp_2.bin.zip](https://github.com/user-attachments/files/18368441/fwbackup_bedside_lamp_2.bin.zip)

just adding the original firmware in case it could be useful

**Comment by ArtikusHG** [comment]
> and could this issue be related to the problem that if I set yellow color in home assistant the lamp outputs something between yellow and green .... other colors look good.

absolutely! my lamp does exactly that and i came here to see if there's a possible solution. this bothers me extremely much as i cannot set the color temperature any warmer than about 4000k, lower than that it becomes really green and looks awful. i really hope we can at least get some options for manual calibration

**Comment by ArtikusHG** [comment]
i was able to partially fix the issue by manually editing the color values in `components/xiaomi_bslamp2/light/color_handler_color_temperature.h`. i did not do anything special, just figured out what my eyes feel like looks better, and ended up with this:

```c++
static const RGBWLevelsTable rgbw_levels_1_ {{
    { 501.0f, 0.673f, 0.907f, 1.000f, 0.063f },
    { 455.0f, 0.673f, 0.896f, 1.000f, 0.063f },
    { 417.0f, 0.673f, 0.891f, 1.000f, 0.068f },
    { 371.0f, 0.673f, 0.880f, 1.000f, 0.070f },
    { 334.0f, 0.673f, 0.887f, 1.000f, 0.088f },
    { 313.0f, 0.682f, 0.904f, 1.000f, 0.128f },
    { 295.0f, 0.947f, 1.000f, 0.968f, 0.145f },
    { 251.0f, 0.999f, 1.000f, 1.000f, 0.155f },
    { 223.0f, 1.000f, 0.899f, 0.921f, 0.130f },
    { 201.0f, 1.000f, 0.873f, 0.908f, 0.115f },
    { 182.0f, 1.000f, 0.873f, 0.901f, 0.103f },
    { 173.0f, 1.000f, 0.873f, 0.904f, 0.094f },
    { 167.0f, 1.000f, 0.873f, 0.891f, 0.098f },
    { 154.0f, 1.000f, 0.873f, 0.894f, 0.090f },
    { 153.0f, 1.000f, 0.873f, 0.892f, 0.088f }
}};

static const RGBWLevelsTable rgbw_levels_100_ {{
    { 501.0f, 0.000f, 0.744f, 1.000f, 0.168f },
    { 455.0f, 0.000f, 0.637f, 1.000f, 0.293f },
    { 417.0f, 0.000f, 0.586f, 1.000f, 0.320f },
    { 371.0f, 0.000f, 0.549f, 1.000f, 0.467f },
    { 334.0f, 0.000f, 0.535f, 1.000f, 0.525f },
    { 313.0f, 0.097f, 0.714f, 1.000f, 0.740f },
    { 295.0f, 0.545f, 1.000f, 0.953f, 0.905f },
    { 251.0f, 1.000f, 1.000f, 1.000f, 1.000f },
    { 223.0f, 1.000f, 0.267f, 0.485f, 0.765f },
    { 201.0f, 1.000f, 0.000f, 0.355f, 0.609f },
    { 182.0f, 1.000f, 0.000f, 0.282f, 0.489f },
    { 173.0f, 1.000f, 0.000f, 0.313f, 0.392f },
    { 167.0f, 1.000f, 0.000f, 0.180f, 0.422f },
    { 154.0f, 1.000f, 0.000f, 0.218f, 0.368f },
    { 153.0f, 1.000f, 0.000f, 0.187f, 0.335f }
}};
```

**note to anyone who will attempt to do this: increasing the value of a channel makes it DIMMER! and vice versa, decreasing the value increases the output of that channel)**

as you can see, all i did was reduce the levels of the green channel. the result is that the light looks much warmer now, and although it's still not very similar to the original firmware, it's much closer and now it doesn't bother me as much.

i originally planned to implement a pr with basic per-channel color calibration, but realized the values don't really scale linearly, so i'm not sure how to approach this. @mmakaay if you have anything in mind, point me to it, and i'll try to implement something in my free time if you don't have time to work on this project.

what would be ideal is, of course, to extract the color values from the original firmware, however, this is a huge undertaking and would probably require a lot of reverse engineering. what i think would be a good and actionable path forward is to make the process i described above easier by implementing a built-in method of color calibration, exposed in the yaml config, and documenting this in the wiki. again, i'm willing to test anything and help work on this, because this issue has been bothering me for months at this point, and since i use adaptive lighting in home assistant to synchronize multiple lights with sunlight during the day, you can imagine how much i hated this :p

**Comment by mmakaay** [comment]
Last week, I have tried calibration ideas, but the enormously non-linear nature of the LED output levels is working against us quite a bit here :-/ 

I even threw some AI at it, to see if it could find secret computations for mapping RGBWW values to the output values to use. I got many happy reports about finding working computations that resulted in output values that differed max 2% with my output level tables. That was all massive hallucination though. Sure, the first three values were correct, but I got to 2500%+ differences when going further up the RGBWW value scales :-D

I can take a look at the original firmware to see if there's anything to reverse engineer from there.

**Comment by ArtikusHG** [comment]
@mmakaay any news on this?

**Comment by mmakaay** [comment]
I've done a few more attempts at disecting the original firmware in the meanwhile, and although it was an interesting exercise, it did not lead to a full understanding of how it drives the colors exactly.  

For the night light in white light mode, a release was just created that can be used to calibrate that specific mode of operation. This is the simplest one, since there is only 1 color to be driven by the code, and that color is now tweakable.
https://github.com/mmakaay/esphome-xiaomi_bslamp2/releases/tag/2026.5.0

Extending that to calibration for all colors is not something I see happening soon, given the non-linear and somewhat strange output levels that I found in the hardware-level reverse engineering. We can't just shift RGB levels a bit for example, since in some cases that leads to very different colors. A linear shift does not result in a linear color change.

I'm a bit out of ideas on this one, but other ideas are welcome of course.

**Comment by rkoptev** [comment]
Any updates on the issue?

**Comment by mmakaay** [comment]
The last update still stands. We have a calibration feature for the night light. For the warm white light difference, I have no ideas how to change the behavior in a generic way that works on all devices.

**Comment by rkoptev** [comment]
Maybe I'm doing something wrong, but no matter how I adjust the 3 settings for the night light (level, temperature, tint) - light color and temperature stays the same. I tried using a calibration table from @ArtikusHG - but it yields reddish and still too bright night light. What I have on default calibration is exactly what @helgek showed on his pictures. The difference is night and day (yep, a silly pun). It's not a bit off, it is order of magnitude brighter, and more like 4000K on the minimal setting compared to the original.
